# Day 4: Session 4A - Missing, Duplicated, Miscast

[Session Webpage](https://eds-217-essential-python.github.io/course-materials/interactive-sessions/4a_cleaning_data.html)

Date: 09/03/2026

In [1]:
import pandas as pd

url = 'https://eds-217-essential-python.github.io/data/marine_microplastics.csv'
plastics = pd.read_csv(url)
plastics.head()

,OBJECTID,Oceans,Regions,SubRegions,Sampling Method,Measurement,Unit,Density Range,Density Class,Short Reference,...,Organization,Keywords,Accession Number,Accession Link,Latitude,Longitude,Date,GlobalID,x,y
0,10008,Atlantic Ocean,NaN,NaN,Grab sample,0.020000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-58.428300,-64.1640,2/3/2017 12:00:00 AM,1e5b8e71-037b-4887-a276-f1e4552acb1f,-64.1640,-58.428300
1,8680,Atlantic Ocean,NaN,NaN,Grab sample,0.008000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-51.308200,-60.5467,11/17/2013 12:00:00 AM,a40f7f7c-1025-4aac-ad16-ee4cba196870,-60.5467,-51.308200
2,13257,Pacific Ocean,NaN,NaN,Manta net,0.019886,pieces/m3,0.005-1,Medium,Faure et al.2015,...,"Oceaneye Association, Switzerland",Oceaneye Association; Citizen Science,276422,https://www.ncei.noaa.gov/access/metadata/land...,-51.826667,-72.5750,12/26/2015 12:00:00 AM,febf79b8-7e2c-46e6-bc15-e08492ec2029,-72.5750,-51.826667
3,9676,Atlantic Ocean,NaN,NaN,Grab sample,0.018000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-31.696000,-48.5600,8/11/2015 12:00:00 AM,a77121b2-e113-444e-82d9-7af11d62fdd2,-48.5600,-31.696000
4,6427,Pacific Ocean,NaN,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2014,...,Sea Education Association,SEA,211008,https://www.ncei.noaa.gov/access/metadata/land...,6.350000,-121.8500,12/18/2002 12:00:00 AM,be27c450-02ca-4261-8d89-cae21108e6cc,-121.8500,6.350000


In [2]:
sea = plastics[plastics['Organization'] == "Sea Education Association"]
sea

,OBJECTID,Oceans,Regions,SubRegions,Sampling Method,Measurement,Unit,Density Range,Density Class,Short Reference,...,Organization,Keywords,Accession Number,Accession Link,Latitude,Longitude,Date,GlobalID,x,y
4,6427,Pacific Ocean,NaN,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2014,...,Sea Education Association,SEA,211008,https://www.ncei.noaa.gov/access/metadata/land...,6.35,-121.85,12/18/2002 12:00:00 AM,be27c450-02ca-4261-8d89-cae21108e6cc,-121.85,6.35
8,2514,Atlantic Ocean,NaN,NaN,Neuston net,0.006480,pieces/m3,0.005-1,Medium,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,40.21,-68.10,7/20/1989 12:00:00 AM,89e16a0e-444c-4d8e-97d3-630f168805bd,-68.10,40.21
9,190,Atlantic Ocean,NaN,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,37.47,-66.97,10/14/1991 12:00:00 AM,c49518f5-74e6-4bba-bf7b-53a133d39bac,-66.97,37.47
11,2111,Atlantic Ocean,NaN,NaN,Neuston net,0.009968,pieces/m3,0.005-1,Medium,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,43.11,-64.37,6/9/2001 12:00:00 AM,80c33a21-2854-4e98-a104-19c362fbea02,-64.37,43.11
12,5431,Atlantic Ocean,NaN,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,43.61,-67.00,6/6/2005 12:00:00 AM,1d53e726-69bb-4bcb-9809-58dc1499be5a,-67.00,43.61
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16229,2874,Atlantic Ocean,NaN,NaN,Neuston net,0.006480,pieces/m3,0.005-1,Medium,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,43.89,-64.40,5/20/1992 12:00:00 AM,098fab78-f883-4b97-bece-5ab3b18a93e0,-64.40,43.89
16231,4234,Atlantic Ocean,Caribbean Sea,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,13.12,-69.81,12/17/1998 12:00:00 AM,e2cd5112-4a84-4328-9399-03e26ab76c6e,-69.81,13.12
16235,1065,Atlantic Ocean,Caribbean Sea,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,12.52,-68.68,12/21/1995 12:00:00 AM,0b37de95-fe6e-4d36-9388-1a08178a2131,-68.68,12.52
16242,4683,Atlantic Ocean,Caribbean Sea,NaN,Neuston net,0.021600,pieces/m3,0.005-1,Medium,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,21.39,-81.96,3/14/2001 12:00:00 AM,83a8208e-2d06-4ad9-bcd9-81e29d57e0aa,-81.96,21.39


In [3]:
plastics.columns.tolist()

['OBJECTID',
 'Oceans',
 'Regions',
 'SubRegions',
 'Sampling Method',
 'Measurement',
 'Unit',
 'Density Range',
 'Density Class',
 'Short Reference',
 'Long Reference',
 'DOI',
 'Organization',
 'Keywords',
 'Accession Number',
 'Accession Link',
 'Latitude',
 'Longitude',
 'Date',
 'GlobalID',
 'x',
 'y']

In [4]:
plastics.isnull().sum()
#  .isnull() returns a True/False for every cell, 
# and .sum() counts the Trues, one column at a time

OBJECTID                0
Oceans                271
Regions              8249
SubRegions          15657
Sampling Method         0
Measurement          5792
Unit                    0
Density Range           0
Density Class           0
Short Reference         0
Long Reference          0
DOI                     0
Organization            0
Keywords               18
Accession Number        0
Accession Link          0
Latitude                0
Longitude               0
Date                    0
GlobalID                0
x                       0
y                       0
dtype: int64

In [5]:
plastics.isnull().sum().sort_values(ascending=False).head(5)

SubRegions     15657
Regions         8249
Measurement     5792
Oceans           271
Keywords          18
dtype: int64

In [6]:
## .dropna()
plastics.dropna().shape
# Bare .dropna() is almost always the wrong tool
# With no arguments, .dropna() removes every row that has a missing value in any column at all

(588, 22)

In [7]:
# subset= takes a list of column names, 
# and only a gap in one of those columns can delete a row:
measured = plastics.dropna(subset=['Measurement'])
measured.shape

(10453, 22)

In [8]:
plastics['Unit'].value_counts()

Unit
pieces/m3           10178
pieces/10 mins       5792
pieces kg-1 d.w.      275
Name: count, dtype: int64

In [9]:
measured['Unit'].value_counts()

Unit
pieces/m3           10178
pieces kg-1 d.w.      275
Name: count, dtype: int64

In [10]:
## Filling gaps with .fillna()
# .fillna() replaces the gaps with a value you choose
plastics['Regions'] = plastics['Regions'].fillna('Unspecified')
plastics['Regions'].value_counts().head(5)

## Kelly Caylor ""
# filling any dataset with new data should be doe with the upmost consideration and care
# dropna and fillna and are "powerful" methods 
# that should only be used when you know EXACTLY what youre doing

Regions
Unspecified          8249
Gulf of Mexico       4817
Caribbean Sea        1886
Mediterranean Sea     345
North Sea             252
Name: count, dtype: int64

In [11]:
plastics.isnull().sum().sort_values(ascending=False).head(3)


SubRegions     15657
Measurement     5792
Oceans           271
dtype: int64

In [12]:
# The other classic defect is the same sample entered twice
# .duplicated() returns a mask, one True or False per row, and .sum() counts them
plastics.duplicated().sum()

0

In [13]:
plastics[['OBJECTID', 'GlobalID', 'Latitude', 'Longitude', 'Date', 'Measurement']].head(3)
# OBJECTID and GlobalID are identifiers, assigned one per record when the archive was built. 
# No two rows can ever share them, so no two rows can ever be identical, 
# so .duplicated() can only ever return zero.
# Any table with an auto-generated ID column will report zero duplicates forever.

,OBJECTID,GlobalID,Latitude,Longitude,Date,Measurement
0,10008,1e5b8e71-037b-4887-a276-f1e4552acb1f,-58.428300,-64.1640,2/3/2017 12:00:00 AM,0.020000
1,8680,a40f7f7c-1025-4aac-ad16-ee4cba196870,-51.308200,-60.5467,11/17/2013 12:00:00 AM,0.008000
2,13257,febf79b8-7e2c-46e6-bc15-e08492ec2029,-51.826667,-72.5750,12/26/2015 12:00:00 AM,0.019886


In [14]:
# need to use "subset=" to point duplicate towards the right columns
plastics.duplicated(subset=sample_columns).sum()
# .duplicated() is only as good as the columns you point it at

NameError: name 'sample_columns' is not defined

### .duplicated() (and realted methods) can be used to create a filter that you then apply to the dataset

In [ ]:
# Define desired columns 
sample_columns = ['Latitude', 'Longitude', 'Date', 'Measurement', 'Unit']

# Create a filter (True/False) using these columns
my_filter = plastics.duplicated(subset=sample_columns, keep=False)

plastics[my_filter]

,OBJECTID,Oceans,Regions,SubRegions,Sampling Method,Measurement,Unit,Density Range,Density Class,Short Reference,...,Organization,Keywords,Accession Number,Accession Link,Latitude,Longitude,Date,GlobalID,x,y
6,14917,Atlantic Ocean,Unspecified,NaN,Megacorer,0.000000,pieces kg-1 d.w.,0-2,Very Low,Courtene-Jones et al. 2020,...,"Scottish Marine Institute, Scotland, United Ki...","Rockall Trough, Scotland",278002,https://www.ncei.noaa.gov/access/metadata/land...,57.298000,-10.374833,5/18/2017 12:00:00 AM,d0cc8862-2466-431e-83ac-96603f4faf78,-10.374833,57.298000
37,15249,Atlantic Ocean,Gulf of Mexico,NaN,PVC cylinder,2115.655853,pieces/m3,500-30000,Medium,Alvarez-Zeferino et al. 2020,...,"Universidad Autónoma de Baja California, Mexico",Mexico,278270,https://www.ncei.noaa.gov/access/metadata/land...,20.974583,-97.306900,4/12/2018 12:00:00 AM,28614d4b-c59f-4a38-80e8-7abb8f1e4300,-97.306900,20.974583
40,21415,Atlantic Ocean,Gulf of Mexico,NaN,Hand picking,NaN,pieces/10 mins,2-40,Medium,Tunnell et al. 2020,...,University of Texas Marine Science Institute,Nurdle Patrol,259486,https://www.ncei.noaa.gov/access/metadata/land...,29.335900,-94.735400,5/30/2022 12:00:00 AM,a284f6a1-6223-411a-956d-12d11d496f40,-94.735400,29.335900
46,18830,Atlantic Ocean,Gulf of Mexico,NaN,Hand picking,NaN,pieces/10 mins,0,Very Low,Tunnell et al. 2020,...,University of Texas Marine Science Institute,Nurdle Patrol,259486,https://www.ncei.noaa.gov/access/metadata/land...,27.851600,-82.554500,12/4/2020 12:00:00 AM,3b3b6417-d3dc-4ca2-818c-105450977f15,-82.554500,27.851600
52,2443,Atlantic Ocean,Caribbean Sea,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,11.940000,-68.260000,12/15/1987 12:00:00 AM,5d35ecdd-8fbf-4613-85ef-2d17d966a062,-68.260000,11.940000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16193,21105,Atlantic Ocean,Gulf of Mexico,NaN,Hand picking,NaN,pieces/10 mins,2-40,Medium,Tunnell et al. 2020,...,University of Texas Marine Science Institute,Nurdle Patrol,259486,https://www.ncei.noaa.gov/access/metadata/land...,29.215600,-94.914500,5/2/2022 12:00:00 AM,90b2e6d6-a395-4107-80a2-bce55d5c60b3,-94.914500,29.215600
16205,15512,Pacific Ocean,Gulf of California,NaN,PVC cylinder,705.218618,pieces/m3,500-30000,Medium,Alvarez-Zeferino et al. 2020,...,"Universidad Autónoma de Baja California, Mexico",Mexico,278270,https://www.ncei.noaa.gov/access/metadata/land...,26.729392,-111.906818,7/11/2018 12:00:00 AM,b89f4701-593b-4195-96b3-55e8b00627dd,-111.906818,26.729392
16208,1386,Atlantic Ocean,Unspecified,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,24.180000,-81.000000,1/8/1998 12:00:00 AM,7228aee5-d08d-42b8-b111-97f0abfbd853,-81.000000,24.180000
16222,17808,Atlantic Ocean,Gulf of Mexico,NaN,Hand picking,NaN,pieces/10 mins,2-40,Medium,Tunnell et al. 2020,...,University of Texas Marine Science Institute,Nurdle Patrol,259486,https://www.ncei.noaa.gov/access/metadata/land...,30.306500,-89.327500,1/24/2020 12:00:00 AM,3988f9bd-491d-4460-91eb-90554d24464b,-89.327500,30.306500


In [ ]:
# .drop_duplicates() takes the same argument and returns a table with the repeats removed, 
# keeping the first of each
deduplicated = plastics.drop_duplicates(subset=sample_columns)
deduplicated.shape

(15389, 22)

In [ ]:
repeats = plastics[plastics.duplicated(subset=sample_columns, keep=False)]
print(repeats.shape)
repeats[['Measurement', 'Unit', 'Sampling Method', 'Short Reference']].head(6)

(1465, 22)


,Measurement,Unit,Sampling Method,Short Reference
6,0.000000,pieces kg-1 d.w.,Megacorer,Courtene-Jones et al. 2020
37,2115.655853,pieces/m3,PVC cylinder,Alvarez-Zeferino et al. 2020
40,NaN,pieces/10 mins,Hand picking,Tunnell et al. 2020
46,NaN,pieces/10 mins,Hand picking,Tunnell et al. 2020
52,0.000000,pieces/m3,Neuston net,Law et al.2010
73,0.002088,pieces/m3,Neuston net,Law et al.2010


In [ ]:
(repeats['Measurement'] == 0).sum()

652

In [ ]:
# dtypes versus astypes

print(plastics.dtypes)
plastics['Accession Number'].head(3)
# showing up as int64 but should be string

OBJECTID              int64
Oceans               object
Regions              object
SubRegions           object
Sampling Method      object
Measurement         float64
Unit                 object
Density Range        object
Density Class        object
Short Reference      object
Long Reference       object
DOI                  object
Organization         object
Keywords             object
Accession Number      int64
Accession Link       object
Latitude            float64
Longitude           float64
Date                 object
GlobalID             object
x                   float64
y                   float64
dtype: object


0    211009
1    211009
2    276422
Name: Accession Number, dtype: int64

In [ ]:
# .astype() converts a column to the type you name
plastics['Accession Number'] = plastics['Accession Number'].astype(str)
# turns "Accession Number" into a string
plastics['Accession Number'].head(3)

0    211009
1    211009
2    276422
Name: Accession Number, dtype: object

In [ ]:
# Easiest object types: .astype(float), .astype(int), and .astype(str)


### Key points

Run df.isnull().sum() first, every time, on any table you did not build yourself.

Bare .dropna() deletes a row for a gap in any column. In this file that cost 96% of the data. Use .dropna(subset=['col']) and name the columns you actually need.

Before you drop, count what you are dropping and look at what it has in common. Missingness usually lines up with something.

.fillna(value) replaces gaps with a value you choose. Choose one that cannot be mistaken for data.


.duplicated() and .drop_duplicates() both take subset=[...]. Without it, any table with an ID column reports zero duplicates forever.

.astype(str/int/float) converts a column. It returns a new column, so assign it back, and remember that narrowing a type throws information away.